File: 01_introduction_to_algorithms/Haskell/01_binarysearch.hs
import Data.Array

binarysearch :: Integer -> Array Integer Integer -> Integer -> Integer -> Maybe Integer
binarysearch x arr low high
  | low > high = Nothing
  | guess > x = binarysearch x arr low (mid - 1)
  | guess < x = binarysearch x arr (mid + 1) high
  | otherwise = Just mid
  where
    mid = ((low + high) `div` 2)
    guess = arr ! mid

find :: Integer -> Array Integer Integer -> Maybe Integer
find x arr = binarysearch x arr low high
  where borders = bounds arr
        low = fst borders
        high = snd borders

-- Usage
-- let arr = array (0, 4) [(i, i*i) | i <- [0..4]]  // [(0,0),(1,1),(2,4),(3,9),(4,16)]
-- find 4 arr                                       // Just 2
-- find 25 arr                                      // Nothing

File: 01_introduction_to_algorithms/lua/01_binary_search.lua
function binary_search(table, value)
    -- Lua arrays start with 1
    local low, high = 1, #table
    
    while low <= high do
        local mid = math.floor((low + high) / 2)
        local guess = table[mid]
        
        if guess == value then
            return mid
        elseif guess > value then
            high = mid - 1
        else
            low = mid + 1
        end
    end
    
    return nil
end

local my_table = {1, 3, 5, 7, 9}

print(binary_search(my_table, 3)) -- => 2
print(binary_search(my_table, -1)) -- => nil